### ISSUE TITLE
Docbook backend applies incorrect column attributes

### ISSUE THREAD

**Issue by mcrotty** [asciidoctor/asciidoctor#1558]
1. Given the following example, the emphasis is not applied correctly once a colspan is involved. (see example 1, row 2, column 3)

.Example (colspan)
[cols="1,1,1e"]
|===
| One column | One column | One column
| Plain      | Plain  | Emphasis  
2+| Plain             | Emphasis  
|===

2. Ditto issue for a rowspan (see example 2, row 4, column 3)

.Example 2 (rowspan  at start)
[cols="1,1,1e"]
|===
| One column | One column | One column
.2+| Plain      | Plain  | Emphasis  
| Plain                       | Emphasis  
|===

3. Ditto issue for a rowspan in middle column (see example 3, row 4, column 3)
.Example 3 (rowspan in middle)
[cols="1,1,1e"]
|===
| One column | One column | One column
| Plain      .2+| Plain  | Emphasis  
| Plain                       | Emphasis  
|===

4. An partial fix is below. Note this is a *partial* fix, because where more than 1 rowspan affects a column, it still can get it wrong. effective_col_visits() counts the number of active rowspans and not the column number where the rowspan impacts.

--- a/lib/asciidoctor/table.rb
+++ b/lib/asciidoctor/table.rb
@@ -473,7 +473,7 @@ class Table::ParserContext
         end
       else
         # QUESTION is this right for cells that span columns?
-        unless (column = @table.columns[@current_row.size])
+        unless (column = @table.columns[effective_col_visits])
           warn %(asciidoctor: ERROR: #{@last_cursor.line_info}: dropping cell because it exceeds specified number of columns)
           return
         end

used effective_col_visits() as it also takes the active rowspans into account.

**Comment by mojavelinux** [comment]
I remember coming across this issue before, but I can't find any reference to related issue. Thanks for shedding light on it again.

The table logic is pretty tricky because it tries to handle all the scenarios with very little code. As long as we have the right number of tests, that's fine. But clearly we're missing a test :) I'll be sure to turn your snippets into tests. Exactly what is needed!

**Comment by mojavelinux** [comment]
The workaround for this issue is to apply the formatting directly to the cell, for those looking for one.

**Comment by mojavelinux** [comment]
This is tangentially related to #989. Maybe the same bit of code, maybe not.

**Comment by mojavelinux** [comment]
Maybe related to #1462 as well?

**Comment by jods4** [comment]
This issue was opened years ago so I'm going to bump it. 👍 
As of asciidoctor.js@1.5.6-preview-1 spans are still messing up cell styles set in `cols`.

**Comment by Mogztter** [comment]
@mojavelinux is this issue related specifically to Asciidoctor.js or to Asciidoctor core (Ruby) ?

**Comment by jods4** [comment]
Given that it was reported in the core repo and that the JS version is an automated port, I would assume it's a core (Ruby) problem?
I can only confirm it reproduces in the JS version as well (I don't use the Ruby version).

**Comment by mojavelinux** [comment]
Yes, this has to be solved in core (Ruby). Rest assured, the issue is scheduled so it's on my radar.

**Comment by danyill** [comment]
#2521 is a duplicate.

Closing in favour of this issue. Only thing there of note is that `activate_rowspans` seems to be much of the cause of complaint. 

A better system for referencing columns from rows is what is required as mentioned above. I see @mcrotty came up with the same (very partial) solution I did. Wish I had found this first! :grin:

**Comment by mojavelinux** [comment]
I'm not surprised that activate_rowspans is the culprit. That method is somewhat of an afterthought given that rowspans were added to the code base later. When I first implemented table parsing, I wasn't planning on supporting row and colspans (because Asciidoctor had much narrower goals back then). So we've probably hit the limit of what that patch could handle and need to apply a more stable strategy.

**Comment by danyill** [comment]
It's a very attractive target... The problem is obvious, the solution less so.

(Also, just so I am clear, `activate_rowspans` is not a culprit with the present behaviour. The present behaviour doesn't even try to accomodate rowspans or colspans for column attributes AFAIK. However to improve it using `effective_col_visits` is  problematic I think because of `activate_rowspans`)

I thought that'll just be an "off by one" thing :laughing: As you say the code is dense because it does a lot while trying to remain performant and compact.
